File: src/main/java/kirjanpito/ui/javafx/EntryTemplateRowModel.java
package kirjanpito.ui.javafx;

import javafx.beans.property.*;
import kirjanpito.db.Account;
import kirjanpito.db.EntryTemplate;

import java.math.BigDecimal;
import java.text.DecimalFormat;

/**
 * JavaFX Properties -pohjainen malli vientimallin riville TableView:ssä.
 */
public class EntryTemplateRowModel {
    private final ObjectProperty<EntryTemplate> template;
    private final IntegerProperty number;
    private final StringProperty name;
    private final ObjectProperty<Account> account;
    private final StringProperty accountName;
    private final ObjectProperty<BigDecimal> debit;
    private final ObjectProperty<BigDecimal> credit;
    private final StringProperty debitFormatted;
    private final StringProperty creditFormatted;
    private final StringProperty description;
    private final DecimalFormat currencyFormat;

    public EntryTemplateRowModel(EntryTemplate template, Account account, DecimalFormat format) {
        this.template = new SimpleObjectProperty<>(template);
        this.currencyFormat = format;

        this.number = new SimpleIntegerProperty(template != null ? template.getNumber() : 1);
        this.name = new SimpleStringProperty(template != null ? template.getName() : "");
        this.account = new SimpleObjectProperty<>(account);
        this.accountName = new SimpleStringProperty(account != null ? account.getNumber() + " " + account.getName() : "");
        this.description = new SimpleStringProperty(template != null ? template.getDescription() : "");

        if (template != null) {
            this.debit = new SimpleObjectProperty<>(template.isDebit() ? template.getAmount() : null);
            this.credit = new SimpleObjectProperty<>(!template.isDebit() ? template.getAmount() : null);
        } else {
            this.debit = new SimpleObjectProperty<>(null);
            this.credit = new SimpleObjectProperty<>(null);
        }

        this.debitFormatted = new SimpleStringProperty(formatAmount(this.debit.get()));
        this.creditFormatted = new SimpleStringProperty(formatAmount(this.credit.get()));

        // Listen for changes
        this.debit.addListener((obs, oldVal, newVal) -> {
            debitFormatted.set(formatAmount(newVal));
            if (this.template.get() != null) {
                this.template.get().setDebit(newVal != null);
                this.template.get().setAmount(newVal != null ? newVal : BigDecimal.ZERO);
            }
        });

        this.credit.addListener((obs, oldVal, newVal) -> {
            creditFormatted.set(formatAmount(newVal));
            if (this.template.get() != null) {
                this.template.get().setDebit(newVal == null);
                this.template.get().setAmount(newVal != null ? newVal : BigDecimal.ZERO);
            }
        });

        this.account.addListener((obs, oldVal, newVal) -> {
            if (newVal != null && this.template.get() != null) {
                this.template.get().setAccountId(newVal.getId());
                accountName.set(newVal.getNumber() + " " + newVal.getName());
            } else if (this.template.get() != null) {
                this.template.get().setAccountId(-1);
                accountName.set("");
            }
        });

        this.number.addListener((obs, oldVal, newVal) -> {
            if (this.template.get() != null) {
                this.template.get().setNumber(newVal.intValue());
            }
        });

        this.name.addListener((obs, oldVal, newVal) -> {
            if (this.template.get() != null) {
                this.template.get().setName(newVal);
            }
        });

        this.description.addListener((obs, oldVal, newVal) -> {
            if (this.template.get() != null) {
                this.template.get().setDescription(newVal);
            }
        });
    }

    private String formatAmount(BigDecimal amount) {
        return amount != null && amount.compareTo(BigDecimal.ZERO) != 0 ? currencyFormat.format(amount) : "";
    }

    // Getters for properties
    public ObjectProperty<EntryTemplate> templateProperty() { return template; }
    public IntegerProperty numberProperty() { return number; }
    public StringProperty nameProperty() { return name; }
    public ObjectProperty<Account> accountProperty() { return account; }
    public StringProperty accountNameProperty() { return accountName; }
    public ObjectProperty<BigDecimal> debitProperty() { return debit; }
    public ObjectProperty<BigDecimal> creditProperty() { return credit; }
    public StringProperty debitFormattedProperty() { return debitFormatted; }
    public StringProperty creditFormattedProperty() { return creditFormatted; }
    public StringProperty descriptionProperty() { return description; }

    // Getters for values
    public EntryTemplate getTemplate() { return template.get(); }
    public int getNumber() { return number.get(); }
    public String getName() { return name.get(); }
    public Account getAccount() { return account.get(); }
    public String getAccountName() { return accountName.get(); }
    public BigDecimal getDebit() { return debit.get(); }
    public BigDecimal getCredit() { return credit.get(); }
    public String getDebitFormatted() { return debitFormatted.get(); }
    public String getCreditFormatted() { return creditFormatted.get(); }
    public String getDescription() { return description.get(); }

    // Setters
    public void setNumber(int number) { this.number.set(number); }
    public void setName(String name) { this.name.set(name); }
    public void setAccount(Account account) { this.account.set(account); }
    public void setDebit(BigDecimal debit) { this.debit.set(debit); }
    public void setCredit(BigDecimal credit) { this.credit.set(credit); }
    public void setDescription(String description) { this.description.set(description); }
}

File: src/main/kotlin/kirjanpito/ui/javafx/EntryTemplateRowModel.kt
package kirjanpito.ui.javafx

import javafx.beans.property.*
import kirjanpito.db.Account
import kirjanpito.db.EntryTemplate
import java.math.BigDecimal
import java.text.DecimalFormat

/**
 * JavaFX Properties -pohjainen malli vientimallin riville TableView:ssä.
 */
class EntryTemplateRowModel(
    template: EntryTemplate?,
    account: Account?,
    format: DecimalFormat
) {
    private val templateProperty = SimpleObjectProperty(template)
    private val number = SimpleIntegerProperty(template?.number ?: 1)
    private val name = SimpleStringProperty(template?.name ?: "")
    private val accountProperty = SimpleObjectProperty(account)
    private val accountName = SimpleStringProperty(
        account?.let { "${it.number} ${it.name}" } ?: ""
    )
    private val debit = SimpleObjectProperty<BigDecimal>(
        template?.takeIf { it.isDebit }?.amount
    )
    private val credit = SimpleObjectProperty<BigDecimal>(
        template?.takeIf { !it.isDebit }?.amount
    )
    private val debitFormatted = SimpleStringProperty(formatAmount(debit.get()))
    private val creditFormatted = SimpleStringProperty(formatAmount(credit.get()))
    private val description = SimpleStringProperty(template?.description ?: "")
    private val currencyFormat = format

    init {
        // Listen for changes
        debit.addListener { _, _, newVal ->
            debitFormatted.set(formatAmount(newVal))
            templateProperty.get()?.let { t ->
                t.setDebit(newVal != null)
                t.amount = newVal ?: BigDecimal.ZERO
            }
        }

        credit.addListener { _, _, newVal ->
            creditFormatted.set(formatAmount(newVal))
            templateProperty.get()?.let { t ->
                t.setDebit(newVal == null)
                t.amount = newVal ?: BigDecimal.ZERO
            }
        }

        accountProperty.addListener { _, _, newVal ->
            templateProperty.get()?.let { t ->
                if (newVal != null) {
                    t.accountId = newVal.id
                    accountName.set("${newVal.number} ${newVal.name}")
                } else {
                    t.accountId = -1
                    accountName.set("")
                }
            }
        }

        number.addListener { _, _, newVal ->
            templateProperty.get()?.number = newVal.toInt()
        }

        name.addListener { _, _, newVal ->
            templateProperty.get()?.name = newVal
        }

        description.addListener { _, _, newVal ->
            templateProperty.get()?.description = newVal
        }
    }

    private fun formatAmount(amount: BigDecimal?): String {
        return if (amount != null && amount.compareTo(BigDecimal.ZERO) != 0) {
            currencyFormat.format(amount)
        } else {
            ""
        }
    }

    // Getters for properties
    fun templateProperty() = templateProperty
    fun numberProperty() = number
    fun nameProperty() = name
    fun accountProperty() = accountProperty
    fun accountNameProperty() = accountName
    fun debitProperty() = debit
    fun creditProperty() = credit
    fun debitFormattedProperty() = debitFormatted
    fun creditFormattedProperty() = creditFormatted
    fun descriptionProperty() = description

    // Getters for values
    fun getTemplate() = templateProperty.get()
    fun getNumber() = number.get()
    fun getName() = name.get()
    fun getAccount() = accountProperty.get()
    fun getAccountName() = accountName.get()
    fun getDebit() = debit.get()
    fun getCredit() = credit.get()
    fun getDebitFormatted() = debitFormatted.get()
    fun getCreditFormatted() = creditFormatted.get()
    fun getDescription() = description.get()

    // Setters
    fun setNumber(number: Int) = this.number.set(number)
    fun setName(name: String) = this.name.set(name)
    fun setAccount(account: Account?) = accountProperty.set(account)
    fun setDebit(debit: BigDecimal?) = this.debit.set(debit)
    fun setCredit(credit: BigDecimal?) = this.credit.set(credit)
    fun setDescription(description: String) = this.description.set(description)
}
